File: src/token.ts
export enum TokenKind {
  Illegal = 'Illegal',
  EOF = 'EOF',

  Ident = 'Ident',
  Int = 'Int',

  Assign = '=',
  Plus = '+',
  Minus = '-',
  Bang = '!',
  Asterisk = '*',
  Slash = '/',

  LT = '<',
  GT = '>',
  Equal = '==',
  NotEqual = '!=',

  Comma = ',',
  Semicolon = ';',

  LParen = '(',
  RParen = ')',
  LBrace = '{',
  RBrace = '}',

  Function = 'Function',
  Let = 'Let',
  If = 'If',
  Else = 'Else',
  Return = 'Return',
  True = 'True',
  False = 'False',
}

const Keywords = {
  fn: TokenKind.Function,
  let: TokenKind.Let,
  if: TokenKind.If,
  else: TokenKind.Else,
  true: TokenKind.True,
  false: TokenKind.False,
  return: TokenKind.Return,
} as const;

export type Token = {
  kind: TokenKind;
  literal: string;
};

export function createToken(type: TokenKind, literal: string): Token {
  return { kind: type, literal };
}

export function lookupIdent(ident: string): TokenKind {
  if (ident in Keywords) {
    return Keywords[ident as keyof typeof Keywords];
  }
  return TokenKind.Ident;
}

const a = 'a'.charCodeAt(0);
const z = 'z'.charCodeAt(0);
const A = 'A'.charCodeAt(0);
const Z = 'Z'.charCodeAt(0);
const _ = '_'.charCodeAt(0);
const _0 = '0'.charCodeAt(0);
const _9 = '9'.charCodeAt(0);

export function isLetter(character: string): boolean {
  const char = character.charCodeAt(0);
  return (a <= char && z >= char) || (A <= char && Z >= char) || char === _;
}

export function isWhitespace(character: string): boolean {
  return character === ' ' || character === '\t' || character === '\n' || character === '\r';
}

export function isDigit(character: string): boolean {
  const char = character.charCodeAt(0);
  return _0 <= char && char <= _9;
}

File: token/token.go
package token

type TokenType string

type Token struct {
	Type    TokenType
	Literal string
}

const (
	ILLEGAL = "ILLEGAL"
	EOF     = "\x00"

	// Identifier + Literal
	IDENT = "IDENT"
	INT   = "INT"

	// Operators
	ASSIGN   = "="
	PLUS     = "+"
	MINUS    = "-"
	ASTERISK = "*"
	SLASH    = "/"
	BANG     = "!"

	LT = "<"
	GT = ">"

	// Delimiters
	COMMA     = ","
	SEMICOLON = ";"

	LPAREN = "("
	RPAREN = ")"
	LBRACE = "{"
	RBRACE = "}"

	// Keywords
	FUNCTION = "fn"
	LET      = "let"
	IF       = "if"
	ELSE     = "else"
	RETURN   = "return"
	TRUE     = "true"
	FALSE    = "false"
)

var keywords = map[string]TokenType{
	"fn":     FUNCTION,
	"let":    LET,
	"if":     IF,
	"else":   ELSE,
	"return": RETURN,
	"true":   TRUE,
	"false":  FALSE,
}

func LookupIdent(ident string) TokenType {
	if tkn, ok := keywords[ident]; ok {
		return tkn
	}
	return IDENT
}
